### PR TITLE
[WIP] Improve DB management

### PR DESCRIPTION
**Pull request by root-beer**
## Update process

`bundle install`

#### /etc/nginx/conf.d/skyhopper.conf
```diff
server {
+   client_max_body_size 1g;  # for uploading database.zip
    ...
}
```

### CODE REVIEW

File: app/controllers/databases_controller.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+class DatabasesController < ApplicationController
+  before_action do
+    authenticate_user! if AppSetting.set?
+  end
+
+  # POST /databases/export
+  def export
+    time = Time.now.strftime('%Y%m%d%H%M%S')
+    zipfile = DatabaseManager.export_as_zip
+    send_file(zipfile.path, filename: "SkyHopper-db-#{Rails.env}-#{time}.zip")
```
Comment (skyarch-networks) [inline app/controllers/databases_controller.rb]: __[brakeman]__
* Model attribute used in file name [:ref](http://brakemanscanner.org/docs/warning_types/file_access/)

File: app/models/database_manager.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+# Copyright (c) 2013-2016 SKYARCH NETWORKS INC.
+#
+# This software is released under the MIT License.
+#
+# http://opensource.org/licenses/mit-license.php
+#
+
+class DatabaseManager
+  SQLPATH = 'tmp/import.sql'
+  SECRETS = [:secret_key_base, :db_crypt_key]
```
Comment (skyarch-networks) [inline app/models/database_manager.rb]: __[rubocop]__
* Style/MutableConstant: Freeze mutable objects assigned to constants.

**Comment by skyarch-networks** [comment]
Click here if you want to create rubocop's auto-correct Pull Request 
[Auto Correct by SideCI](https://sideci.com/gh/31754199/issues/pull_request/344/diff/85a9b11feb302dba5a91ef15b010b8994075e55b?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

File: app/models/database_manager.rb [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      env = SUFFIX[Rails.env]
+      secrets.each do |key, value|
+        path = Rails.root.join('secrets', "#{key}-#{env}")
+        File.write(path, value)
+      end
+
+      ReloadSecretsJob.perform_now   # runs in Rails process
+      ReloadSecretsJob.perform_later # runs in Sidekiq process
+
+      system("rake db:data:load[#{sqlpath}]")
```
Comment (skyarch-networks) [inline app/models/database_manager.rb]: __[brakeman]__
* Possible command injection [:ref](http://brakemanscanner.org/docs/warning_types/command_injection/)

File: app/jobs/reload_secrets_job.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+# http://opensource.org/licenses/mit-license.php
+#
+
+class ReloadSecretsJob < ActiveJob::Base
+  queue_as :default
+
+  # Ref: https://github.com/rails/rails/blob/634741d9721eb938c8bce38c109023178268e43d/railties/lib/rails/application.rb#L385
+  def perform
+    Rails.application.remove_instance_variable(:@secrets)
+  rescue NameError => ex
```
Comment (skyarch-networks) [inline app/jobs/reload_secrets_job.rb]: __[rubocop]__
* Lint/HandleExceptions: Do not suppress exceptions. [:ref](https://github.com/bbatsov/ruby-style-guide#dont-hide-exceptions)
* Lint/UselessAssignment: Useless assignment to variable - `ex`. [:ref](https://github.com/bbatsov/ruby-style-guide#underscore-unused-vars)

**Comment by skyarch-networks** [comment]
Click here if you want to create rubocop's auto-correct Pull Request 
[Auto Correct by SideCI](https://sideci.com/gh/31754199/issues/pull_request/344/diff/75654afb00af22a43374f757ce5d186b6edac458?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

File: app/controllers/databases_controller.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    man = DatabaseManager.new
+    def man.policy_class; DatabasePolicy end
+    authorize man
+  end
+
+  # POST /databases/export
+  def export
+    time = Time.now.strftime('%Y%m%d%H%M%S')
+    zipfile = DatabaseManager.export_as_zip
+    send_file(zipfile.path, filename: "SkyHopper-db-#{Rails.env}-#{time}.zip")
```
Comment (skyarch-networks) [inline app/controllers/databases_controller.rb]: __[brakeman]__
* Model attribute used in file name [:ref](http://brakemanscanner.org/docs/warning_types/file_access/)

File: app/controllers/databases_controller.rb [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+  helper_method :app_setting_is_set
+
+  # POST /databases/export
+  def export
+    authenticate_user!
+
+    time = Time.now.strftime('%Y%m%d%H%M%S')
+    zipfile = DatabaseManager.export_as_zip
+    send_file(zipfile.path, filename: "SkyHopper-db-#{Rails.env}-#{time}.zip")
```
Comment (skyarch-networks) [inline app/controllers/databases_controller.rb]: __[brakeman]__
* Model attribute used in file name [:ref](http://brakemanscanner.org/docs/warning_types/file_access/)

**Comment by root-beer** [comment]
Please review and merge PR. @Joeper214 :bow:

**Comment by Joeper214** [comment]
I'll review now. and Resolve conflicts on Gemfile.lock

**Comment by Joeper214** [comment]
How long does the maintenance takes? I tried to import database but I don't know when will be it finished. I refreshed the browser many times. but still no response.

File: app/assets/javascripts/databases.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+// http://opensource.org/licenses/mit-license.php
+//
+
+(function () {
+  'use strict';
+
+  Vue.component('div-loader', Loader);
+  var modal = require('modal');
+
+  var vm = new Vue({
```
Comment (Joeper214) [inline app/assets/javascripts/databases.js]: I'm a bit confused if this loading feature is being used.

**Comment by Joeper214** [comment]
I have encountered a problem. When importing database, I have found that when reloading the page  immediately, the importing process breaks. And the application will remain on the `turnout` page even after an hour or more.

**Comment by root-beer** [comment]
Sorry, It may be a bug. I'll investigate it.

**Comment by root-beer** [comment]
Import will be completed in a few seconds if it is working properly.

**Comment by Joeper214** [comment]
Is it also possible if we can write test scripts on these?

File: app/models/database_manager.rb [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
```
Comment (skyarch-networks) [inline app/models/database_manager.rb]: __[brakeman]__
* Possible command injection [:ref](http://brakemanscanner.org/docs/warning_types/command_injection/)

File: app/controllers/databases_controller.rb [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
```
Comment (skyarch-networks) [inline app/controllers/databases_controller.rb]: __[brakeman]__
* Model attribute used in file name [:ref](http://brakemanscanner.org/docs/warning_types/file_access/)

File: app/views/databases/_modal_export.html.erb [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+              <li><%= t('databases.msg.secret')%></li>
+              <li><%= t('databases.msg.db_key')%></li>
+            </ul>
+          </div>
+        </div>
+        <form action="<%= export_database_path %>" method="POST">
+          <input type="hidden" name="authenticity_token" value="<%= form_authenticity_token %>" >
+          <button type="submit" class="btn btn-primary btn-lg btn-block" :disabled="loading" @click="">
+            <div-loader v-if="loading"></div-loader>
+            <span v-if="!loading">
```
Comment (Joeper214) [inline app/views/databases/_modal_export.html.erb]: I guess this loading function is not being used?

Comment (root-beer) [inline app/views/databases/_modal_export.html.erb]: Yes. I'll remove it.

**Comment by Joeper214** [comment]
Mind if  I update the English translation?

**Comment by root-beer** [comment]
よろしくお願いします！ :bow:

**Comment by Joeper214** [comment]
No other changes on this PR @root-beer ?

**Comment by root-beer** [comment]
Please wait for a while because we can not guarantee that it works reliably yet.

**Comment by Joeper214** [comment]
Thanks. for making it WIP.

**Comment by root-beer** [comment]
@Joeper214 Could you try this on EC2? :bow:

**Comment by Joeper214** [comment]
Yes I'll try. Thanks.

**Comment by Joeper214** [comment]
Please change WIP status

**Comment by Joeper214** [comment]
LGTM! Thanks. 🙇

**Comment by root-beer** [comment]
Thanks! :bow:

**Pull request merged**